src/sciline/pipeline.py

```diff
@@ -503,12 +532,27 @@ def _get_provider(
         ) is not None:
             requested = get_args(tp)
             matches = [
-                (args, subprovider)
+                (args, subprovider, bound)
                 for args, subprovider in subproviders.items()
-                if _is_compatible_type_tuple(requested, args)
+                if (bound := _is_compatible_type_tuple(requested, args)) is not None
             ]
+            # Determining what provider is most specialized is a bit ambiguous.
+            # Is it the provider with fewest arguments that are not concret types?
+            # Or is it the provider that needs to define the fewest TypeVars
+            # to become concrete?
+            #
+            # Contrived example:
+            #  We are requiring Person[Likes[Green], Green] and find the following
+            #  matching providers to choose from:
+            #    Person[Likes[~Color], ~Color] <- matches if we set ~Color = Green
+            #    Person[~Preference, Green] <- matches if we
+            #                                        set ~Preference = Likes[Green]
+            #
+            # Currently the second will be chosen because only one of the arguments
+            # is not concrete while for the first type both arguments are not concrete.
```

Comment:
Missing a test for this?

Reply:
Yes I'll add a test for this. But I think we also need to decide what behavior we want there. I'm leaning towards changing it to count the number of type-vars in the whole type expression (the length of `bound`). In the above example that would lead to an `AmbiguousProviderError` because both expressions have the same number of type-vars. This seems more appropriate because I don't think that it's obvious that one provider is more specialized than the other.

What do you think?

Reply:
Yes, we should raise.